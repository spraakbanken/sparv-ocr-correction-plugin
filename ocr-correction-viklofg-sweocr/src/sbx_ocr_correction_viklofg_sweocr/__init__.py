from typing import Optional

from sparv.api import (  # type: ignore [import-untyped]
    Annotation,
    Output,
    annotator,
    get_logger,
)
from transformers import (  # type: ignore [import-untyped]
    AutoTokenizer,
    T5ForConditionalGeneration,
    pipeline,
)

__description__ = "Calculating word neighbours by mask a word in a BERT model."

MODEL_NAME = "viklofg/swedish-ocr-correction"
MODEL_REVISION = "84b138048992271be7617ccb11056bbcb9b72262"
TOKENIZER_NAME = "google/byt5-small"
TOKENIZER_REVISION = "68377bdc18a2ffec8a0533fef03b1c513a4dd49d"


__version__ = "0.2.0"

logger = get_logger(__name__)

TOK_SEP = " "


@annotator(
    "Word neighbour tagging with a masked Bert model",
)
def annotate_ocr_correction(
    out_ocr_correction: Output = Output(
        "<token>:sbx_ocr_correction_viklofg_sweocr.ocr-correction--viklofg-sweocr",
        cls="sbx_ocr_correction_viklofg_sweocr",
        description="OCR Corrections from viklfog/swedish-ocr (format: '|<word>:<score>|...|)",  # noqa: E501
    ),
    word: Annotation = Annotation("<token:word>"),
    sentence: Annotation = Annotation("<sentence>"),
) -> None:
    ocr_corrector = OcrCorrector.default()

    sentences, _orphans = sentence.get_children(word)
    token_word = list(word.read())
    out_ocr_correction_annotation = word.create_empty_attribute()

    logger.progress(total=len(sentences))  # type: ignore
    for sent in sentences:
        logger.progress()  # type: ignore
        sent_to_tag = [token_word[token_index] for token_index in sent]

        ocr_corrections = ocr_corrector.calculate_corrections(sent_to_tag)
        out_ocr_correction_annotation[:] = ocr_corrections

    logger.info("writing annotations")
    out_ocr_correction.write(out_ocr_correction_annotation)


class OcrCorrector:
    TEXT_LIMIT: int = 127

    def __init__(self, *, tokenizer, model) -> None:
        self.tokenizer = tokenizer
        self.model = model
        self.pipeline = pipeline(
            "text2text-generation", model=model, tokenizer=tokenizer
        )

    @classmethod
    def default(cls) -> "OcrCorrector":
        tokenizer = AutoTokenizer.from_pretrained(
            TOKENIZER_NAME, revision=TOKENIZER_REVISION
        )
        model = T5ForConditionalGeneration.from_pretrained(
            MODEL_NAME, revision=MODEL_REVISION
        )
        return cls(model=model, tokenizer=tokenizer)

    def calculate_corrections(self, text: list[str]) -> list[Optional[str]]:
        logger.debug("Analyzing '%s'", text)
        parts = []
        curr_part: list[str] = []
        curr_len = 0
        ocr_corrections: list[str] = []
        for word in text:
            len_word = bytes_length(word)
            if (curr_len + len_word + 1) > self.TEXT_LIMIT:
                parts.append(TOK_SEP.join(curr_part))
                curr_part, curr_len = [word], len_word
            else:
                curr_part.append(word)
                curr_len = len_word if curr_len == 0 else curr_len + len_word + 1
        if len(curr_part) > 0:
            parts.append(TOK_SEP.join(curr_part))
        for part in parts:
            suggested_text = self.pipeline(part)[0]["generated_text"]
            suggested_text = suggested_text.replace(",", " ,")
            suggested_text = suggested_text.replace(".", " .")
            ocr_corrections = ocr_corrections + suggested_text.split(TOK_SEP)

        if len(text) == len(ocr_corrections) + 1 and text[-1] != ocr_corrections[-1]:
            ocr_corrections.append(text[-1])
        return zip_and_diff(text, ocr_corrections)


def zip_and_diff(orig: list[str], sugg: list[str]) -> list[Optional[str]]:
    return [sw if sw != ow else None for (ow, sw) in zip(orig, sugg)]


def bytes_length(s: str) -> int:
    return len(s.encode("utf-8"))
