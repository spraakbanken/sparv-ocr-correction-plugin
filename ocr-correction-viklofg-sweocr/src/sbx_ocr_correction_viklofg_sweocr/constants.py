"""Constants."""

PROJECT_NAME: str = "sbx_ocr_correction_viklofg_sweocr"
