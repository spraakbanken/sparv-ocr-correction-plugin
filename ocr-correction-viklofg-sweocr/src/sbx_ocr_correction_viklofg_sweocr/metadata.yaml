id: swe-sbx-ocr-correction-viklofg-sweocr
name:
    swe: OCR korrigering med viklofg/swedish-ocr-correction för svenska
    eng: OCR correction with viklofg/swedish-ocr-correction for Swedish
short_description:
    swe: Annotering av OCR-korrigeringar.
    eng: OCR correction annotations
task: ocr-correction
annotations:
    - sbx_ocr_correction_viklofg_sweocr.sbx-ocr-correction
    - sbx_ocr_correction_viklofg_sweocr.sbx-ocr-correction:sbx_ocr_correction_viklofg_sweocr.ocr-correction--viklofg-sweocr
example_output: |-
    ```xml
    <token pos="KN">Och</token>
    <token pos="HA">då</token>
    <token pos="PN">de</token>
    <token pos="VB">räknade</token>
    <token pos="MID">,</token>
    <sbx-ocr-correction ocr-correction--viklofg-sweocr="så">
      <token pos="AB">si</token>
    </sbx-ocr-correction>
    <token pos="MID">,</token>
    <token pos="AB">då</token>
    ```
