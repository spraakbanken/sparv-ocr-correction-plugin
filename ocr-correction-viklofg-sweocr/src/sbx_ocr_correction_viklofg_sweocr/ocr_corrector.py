import re
from typing import List, Optional, Tuple

from parallel_corpus import graph
from parallel_corpus.token import Token
from sparv import api as sparv_api  # type: ignore [import-untyped]
from transformers import (  # type: ignore [import-untyped]
    AutoTokenizer,
    T5ForConditionalGeneration,
    pipeline,
)


def bytes_length(s: str) -> int:
    return len(s.encode("utf-8"))


TOK_SEP = " "
logger = sparv_api.get_logger(__name__)
TOKENIZER_REVISION = "68377bdc18a2ffec8a0533fef03b1c513a4dd49d"
TOKENIZER_NAME = "google/byt5-small"
MODEL_REVISION = "84b138048992271be7617ccb11056bbcb9b72262"
MODEL_NAME = "viklofg/swedish-ocr-correction"

PUNCTUATION = re.compile(r"[.,:;!?]")


class OcrCorrector:
    TEXT_LIMIT: int = 127

    def __init__(self, *, tokenizer, model) -> None:
        self.tokenizer = tokenizer
        self.model = model
        self.pipeline = pipeline(
            "text2text-generation", model=model, tokenizer=tokenizer
        )

    @classmethod
    def default(cls) -> "OcrCorrector":
        tokenizer = AutoTokenizer.from_pretrained(
            TOKENIZER_NAME, revision=TOKENIZER_REVISION
        )
        model = T5ForConditionalGeneration.from_pretrained(
            MODEL_NAME, revision=MODEL_REVISION
        )
        return cls(model=model, tokenizer=tokenizer)

    def calculate_corrections(
        self, text: List[str]
    ) -> List[Tuple[Tuple[int, int], Optional[str]]]:
        logger.debug("Analyzing '%s'", text)

        parts: List[str] = []
        curr_part: List[str] = []
        curr_len = 0
        ocr_corrections: List[Tuple[Tuple[int, int], Optional[str]]] = []
        for word in text:
            len_word = bytes_length(word)
            if (curr_len + len_word + 1) > self.TEXT_LIMIT:
                parts.append(TOK_SEP.join(curr_part))
                curr_part, curr_len = [word], len_word
            else:
                curr_part.append(word)
                curr_len = len_word if curr_len == 0 else curr_len + len_word + 1
        if len(curr_part) > 0:
            parts.append(TOK_SEP.join(curr_part))
        curr_start = 0
        for part in parts:
            suggested_text = self.pipeline(part)[0]["generated_text"]
            suggested_text = PUNCTUATION.sub(r" \g<0>", suggested_text)
            graph_aligned = graph.init_with_source_and_target(part, suggested_text)
            span_ann, curr_start = align_and_diff(graph_aligned, curr_start=curr_start)
            ocr_corrections.extend(span_ann)

        logger.debug("Finished analyzing. ocr_corrections=%s", ocr_corrections)
        return ocr_corrections


def align_and_diff(
    g: graph.Graph, *, curr_start: int
) -> Tuple[List[Tuple[Tuple[int, int], Optional[str]]], int]:
    corrections = []
    edge_map = graph.edge_map(g)
    visited_tokens = set()
    for s_token in g.source:
        logger.debug("checking s_token=%s", s_token)
        edge = edge_map[s_token.id]

        source_ids = [id_ for id_ in edge.ids if id_.startswith("s")]
        target_ids = [id_ for id_ in edge.ids if id_.startswith("t")]
        target_ids_str = "-".join(target_ids)
        if target_ids_str in visited_tokens:
            continue
        visited_tokens.add(target_ids_str)
        logger.debug("processing s_token=%s", s_token)

        if len(source_ids) == len(target_ids):
            source_text = "".join(
                lookup_text(g.source, s_id) for s_id in source_ids
            ).strip()
            target_text = "".join(
                lookup_text(g.target, s_id) for s_id in target_ids
            ).strip()
            start = curr_start
            curr_start += 1
            corrections.append(
                (
                    (start, curr_start),
                    target_text if source_text != target_text else None,
                )
            )

        elif len(source_ids) == 1:
            target_texts = " ".join(
                lookup_text(g.target, id_).strip() for id_ in target_ids
            )
            source_text = s_token.text.strip()
            start = curr_start
            curr_start += 1

            corrections.append(
                (
                    (start, curr_start),
                    target_texts if source_text != target_texts else None,
                ),
            )
        elif len(target_ids) == 1:
            target_text = lookup_text(g.target, target_ids[0]).strip()
            start = curr_start
            curr_start += len(source_ids)
            corrections.append(((start, curr_start), target_text))
        else:
            # TODO Handle this correct (https://github.com/spraakbanken/sparv-sbx-ocr-correction/issues/50)
            raise NotImplementedError(
                f"Handle several sources, {source_ids=} {target_ids=} {g.source=} {g.target=}"  # noqa: E501
            )

    return corrections, curr_start


def lookup_text(tokens: List[Token], id_: str) -> str:
    for token in tokens:
        if token.id == id_:
            return token.text

    raise ValueError(
        f"The id={id_} isn't found in the list of tokens",
    )
