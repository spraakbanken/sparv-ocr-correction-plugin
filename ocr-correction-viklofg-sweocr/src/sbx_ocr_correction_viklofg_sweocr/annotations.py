from sparv import api as sparv_api  # type: ignore [import-untyped]
from sparv.api import Annotation, Output, annotator  # type: ignore [import-untyped]

from sbx_ocr_correction_viklofg_sweocr.ocr_corrector import OcrCorrector

logger = sparv_api.get_logger(__name__)


@annotator("OCR corrections as annotations", language=["swe"])
def annotate_ocr_correction(
    out_ocr_correction: Output = Output(
        "<token>:sbx_ocr_correction_viklofg_sweocr.ocr-correction--viklofg-sweocr",
        cls="sbx_ocr_correction_viklofg_sweocr",
        description="OCR Corrections from viklfog/swedish-ocr (format: '|<word>:<score>|...|)",  # noqa: E501
    ),
    word: Annotation = Annotation("<token:word>"),
    sentence: Annotation = Annotation("<sentence>"),
) -> None:
    ocr_corrector = OcrCorrector.default()

    sentences, _orphans = sentence.get_children(word)
    token_word = list(word.read())
    out_ocr_correction_annotation = word.create_empty_attribute()

    logger.progress(total=len(sentences))  # type: ignore
    for sent in sentences:
        logger.progress()  # type: ignore
        sent_to_tag = [token_word[token_index] for token_index in sent]

        ocr_corrections = ocr_corrector.calculate_corrections(sent_to_tag)
        for i, ocr_correction in enumerate(ocr_corrections, start=sent[0]):
            out_ocr_correction_annotation[i] = ocr_correction

    logger.info("writing annotations")
    out_ocr_correction.write(out_ocr_correction_annotation)
