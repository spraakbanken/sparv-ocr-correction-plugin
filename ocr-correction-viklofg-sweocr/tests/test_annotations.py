from sbx_ocr_correction_viklofg_sweocr.annotations import annotate_ocr_correction
from sparv_pipeline_testing import MemoryOutput, MockAnnotation


def test_annotate_ocr_correction(snapshot) -> None:  # noqa: ANN001
    output_ocr: MemoryOutput = MemoryOutput()
    output_ocr_corr: MemoryOutput = MemoryOutput()
    # "Jonath an saknades ."
    # "12345678901234567890"
    # "         1         2"
    word = MockAnnotation(name="<token:word>", values=["Jonath", "an", "saknades", "."])
    sentence = MockAnnotation(name="<sentence>", children={"<token:word>": [[0, 1, 2, 3]]})
    token = MockAnnotation(name="<token>", spans=[(0, 6), (7, 9), (10, 18), (19, 20)])

    annotate_ocr_correction(output_ocr, output_ocr_corr, word, sentence, token)

    assert output_ocr.values == snapshot
    assert output_ocr_corr.values == snapshot
