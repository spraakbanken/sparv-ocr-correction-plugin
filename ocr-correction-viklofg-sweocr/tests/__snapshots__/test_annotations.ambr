# serializer version: 1
# name: test_annotate_ocr_correction
  list([
    tuple(
      7,
      18,
    ),
  ])
# ---
# name: test_annotate_ocr_correction.1
  list([
    'ansaknades',
  ])
# ---
